File: algorithms/22_haskell/dp/LCS.hs
module LCS (lcs) where
lcs :: Eq a => [a] -> [a] -> Int
lcs [] _ = 0
lcs _ [] = 0
lcs (x:xs) (y:ys)
  | x==y = 1 + lcs xs ys
  | otherwise = max (lcs xs (y:ys)) (lcs (x:xs) ys)

File: algorithms/15_ruby/dp/lcs.rb
def lcs(a, b)
  m, n = a.length, b.length
  dp = Array.new(m+1) { Array.new(n+1, 0) }
  (1..m).each { |i| (1..n).each { |j|
    dp[i][j] = a[i-1]==b[j-1] ? dp[i-1][j-1]+1 : [dp[i-1][j],dp[i][j-1]].max } }
  dp[m][n]
end